File: 2025/python/part1.py
#!/usr/bin/env python3

from sys import argv

def extract_joltage(line: str):
    last_but_one = len(line) - 1
    max1 = None
    max2 = None
    for idx, ch in enumerate(line):
        if idx < last_but_one:
            if not max1 or ch > max1:
                max1 = ch
                max2 = None
            elif not max2 or ch > max2:
                max2 = ch
        else:
            if not max2 or ch > max2:
                max2 = ch
    assert(max1 and max2)
    return (ord(max1) - ord('0')) * 10 + (ord(max2) - ord('0'))

def main():
    fin = 0 if len(argv) != 2 else argv[1]
    lines = open(fin).read().strip().split("\n")
    result = 0
    for line in lines:
        result += extract_joltage(line)
    print(result)

if __name__ == "__main__":
    main()

File: 2025/rust/src/bin/part1.rs
use aoc_2025_03::pac::StdinOrFile;

fn bank_joltage(line: &str) -> u32 {
    let max1 = line[..line.len() - 1].bytes().max().unwrap();

    let idx1 = line.bytes()
        .enumerate()
        .find(|(_, v)| *v == max1)
        .unwrap().0;

    let max2 = line[idx1 + 1..].bytes().max().unwrap();

    let c1 = max1 as u32 - '0' as u32;
    let c2 = max2 as u32 - '0' as u32;

    c1*10 + c2
}

fn main() {
    let mut result = 0;

    for line in StdinOrFile::lines().map_while(Result::ok) {
        result += bank_joltage(&line);
    }

    println!("{result}");
}
